bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/discovery/MideaACDiscoveryService.java

```diff
@@ -0,0 +1,356 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.mideaac.internal.discovery;
+
+import static org.openhab.binding.mideaac.internal.MideaACBindingConstants.*;
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.TreeMap;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.mideaac.internal.Utils;
+import org.openhab.binding.mideaac.internal.handler.CommandBase;
+import org.openhab.binding.mideaac.internal.security.CloudProvider;
+import org.openhab.binding.mideaac.internal.security.Security;
+import org.openhab.core.config.discovery.AbstractDiscoveryService;
+import org.openhab.core.config.discovery.DiscoveryResult;
+import org.openhab.core.config.discovery.DiscoveryResultBuilder;
+import org.openhab.core.config.discovery.DiscoveryService;
+import org.openhab.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link MideaACDiscoveryService} service for Midea AC.
+ *
+ * @author Jacek Dobrowolski - Initial contribution
+ * @author Bob Eckhoff - OH naming conventions
+ */
+@NonNullByDefault
+@Component(service = DiscoveryService.class, configurationPid = "discovery.mideaac")
+public class MideaACDiscoveryService extends AbstractDiscoveryService {
+
+    private static int discoveryTimeoutSeconds = 5;
+    private final int receiveJobTimeout = 20000;
+    private final int udpPacketTimeout = receiveJobTimeout - 50;
+    private final String mideaacNamePrefix = "MideaAC";
+
+    private final Logger logger = LoggerFactory.getLogger(MideaACDiscoveryService.class);
+
+    ///// Network
+    private byte[] buffer = new byte[512];
+    @Nullable
+    private DatagramSocket discoverSocket;
+
+    @SuppressWarnings("unused")
+    private boolean fullDiscovery = false;
+    @Nullable
+    DiscoveryHandler discoveryHandler;
+
+    private Security security;
+
+    /**
+     * Discovery Service
+     */
+    public MideaACDiscoveryService() {
+        super(SUPPORTED_THING_TYPES_UIDS, discoveryTimeoutSeconds, false);
+        this.security = new Security(CloudProvider.getCloudProvider(""));
+    }
+
+    @Override
+    protected void startScan() {
+        logger.debug("Start scan for Midea AC devices.");
+        discoverThings();
+    }
+
+    @Override
+    protected void stopScan() {
+        logger.debug("Stop scan for Midea AC devices.");
+        closeDiscoverSocket();
+        super.stopScan();
+    }
+
+    /**
+     * Performs the actual discovery of Midea AC devices (things).
+     */
+    @SuppressWarnings("null")
```

Comment:
Please remove all null suppressions.

Reply:
Going to need some guidance here. Is the OH standard (or is it a preference item) to have the Warning during “mvn clean install” and in the code editor (I use VSC) Example from recent compile: (MideaACHandler.java:[1412,31] Potential null pointer access: this expression has a '@Nullable' type)

Tl;dr; One of the main effort in cleaning up the binding for was adding @NonNullByDefault to each class.  This included adding @Nullable to several variables that can be null.  When those variables were used in a method, VSC suggested the quick fix to add @SuppressWarnings("null").  I took that to mean “This (@Nullable) situation has been looked at, so no need to “Warn” during compiling”. This cleared the warnings in VSC while editing, and during “mvn clean install”.  In a couple of cases I still saw “At least one of the problems in category 'null' is not analyzed due to a compiler option being ignored”, but when I removed @SuppressWarnings("null") I got a warning during “mvn clean install”.  What was in this PR was the no Warnings scenario because it looked cleaner to me.

Reply:
Totally understand the struggle. `NullReferenceExceptions` are a common issue and the annotations try to prevend them. The annotations are nothing more then hints to the compiler, the compiled code itself is not changed. The compiler can use these hints to prevent you from creating bugs. E.g. if you annotated some member as `@Non-null` and you assign `null` to it that would be a bug. If you have this in mind, you might also understand that it is not desired to one hand mark annotate and on the other suppress the null warnings.
Ideally we add annotations and handle the null warnings.

Reply:
Don't quite understand, but it worked!  I'll see how far I can get.